src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/VsTelemetryService.cs

```diff
@@ -1,20 +1,142 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
 using System.ComponentModel.Composition;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem;
 
 namespace Microsoft.VisualStudio.Telemetry
 {
     [Export(typeof(ITelemetryService))]
-    internal class VsTelemetryService : ITelemetryService
+    [Export(typeof(IVsTelemetryService))]
+    internal class VsTelemetryService : IVsTelemetryService, ITelemetryService
     {
+        private const string EventPrefix = "vs/projectsystem/managed/";
+        private const string PropertyPrefix = "VS.ProjectSystem.Managed.";
+        private readonly IUnconfiguredProjectCommonServices _unconfiguredProjectCommonServices;
+        private readonly object _lock =  new object();
+
+        private TelemetryEventCorrelation _telemetryEventCorrelation;
+        private bool _telemetryEventCorrelationInitialized;
+        private Guid _projectGuid;
+
+        [ImportingConstructor]
+        public VsTelemetryService(IUnconfiguredProjectCommonServices unconfiguredProjectCommonServices)
+        {
+            Requires.NotNull(unconfiguredProjectCommonServices, nameof(unconfiguredProjectCommonServices));
+
+            _unconfiguredProjectCommonServices = unconfiguredProjectCommonServices;
+        }
+
+        /// <summary>
+        /// Creates the Project Correlation asset which will be used to correlate all the telemetry events for this project
+        /// identified uniquely through the Project Guid.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> CreateProjectCorrelationAssetAsync()
+        {
+            var configurationGeneralProperties = await _unconfiguredProjectCommonServices
+                                                .ActiveConfiguredProjectProperties
+                                                .GetConfigurationGeneralPropertiesAsync()
+                                                .ConfigureAwait(false);
+
+            if (Guid.TryParse((string)await configurationGeneralProperties.ProjectGuid.GetValueAsync().ConfigureAwait(false), out Guid guid))
+            {
+                _projectGuid = guid;
+            }
+            else
+            {
+                return false;
+            }
+
+            var targetFrameworkProperty = await configurationGeneralProperties.TargetFramework.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+            var targetFrameworksProperty = await configurationGeneralProperties.TargetFrameworks.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+            var outputTypeProperty = await configurationGeneralProperties.OutputType.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+
+            lock (_lock)
```

Comment:
This is a bit more complicated.

The creation of the asset can fails if the asset is tried to be created before the ProjectGuid is set. In this case, we will post the telemetry without the asset correlation. The next time we try post the telemetry, we check if the asset was created. If not, we try to create it again. The assumption is, at some point, the ProjectGuid will be set and the asset creation will succeed. The subsequent telemetry data can be correlated to the asset.

If AsyncLazy allows factory that take argument then we try to use it. 

At this point, `lock` is much easier to use and less complicated.